cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisionerStore.java

```diff
@@ -0,0 +1,93 @@
+ */
+
+package co.cask.cdap.internal.provision;
+
+import co.cask.cdap.proto.id.ProgramRunId;
+import co.cask.cdap.spi.data.StructuredTableContext;
+import co.cask.cdap.spi.data.TableNotFoundException;
+import co.cask.cdap.spi.data.transaction.TransactionRunner;
+import co.cask.cdap.spi.data.transaction.TransactionRunners;
+
+import java.io.IOException;
+import java.util.List;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * Stores information used for provisioning.
+ *
+ * Stores subscriber offset information for TMS, cluster information for program runs, and state information for
+ * each provision and deprovision operation.
+ *
+ * Provisioner Store uses transactionRunners to perform underlying CRUD operations.
+ */
+public class ProvisionerStore {
+
+  private final TransactionRunner txRunner;
+
+  private ProvisionerTable getProvisionerTable(StructuredTableContext context) throws TableNotFoundException {
+    return new ProvisionerTable(context);
+  }
+
+  @Inject
+  public ProvisionerStore(TransactionRunner txRunner) {
+    this.txRunner = txRunner;
+  }
+
+  public List<ProvisioningTaskInfo> listTaskInfo() throws IOException {
+    return TransactionRunners.run(txRunner, context -> {
+      return getProvisionerTable(context).listTaskInfo();
+    }, IOException.class);
+  }
+
+  @Nullable
+  public ProvisioningTaskInfo getTaskInfo(final ProvisioningTaskKey key) throws IOException {
+    return TransactionRunners.run(txRunner, context -> {
+      return getProvisionerTable(context).getTaskInfo(key);
+    }, IOException.class);
+  }
+
+  public void putTaskInfo(final ProvisioningTaskInfo taskInfo) throws IOException {
+    TransactionRunners.run(txRunner, context -> {
+      getProvisionerTable(context).putTaskInfo(taskInfo);
+    }, IOException.class);
+  }
+
+  public void deleteTaskInfo(ProgramRunId programRunId) throws IOException {
+    TransactionRunners.run(txRunner, context -> {
+      getProvisionerTable(context).deleteTaskInfo(programRunId);
+    }, IOException.class);
+  }
+
+  @Nullable
+  public ProvisioningTaskInfo getExistingAndCancel(final ProvisioningTaskKey taskKey) throws IOException {
+    return TransactionRunners.run(txRunner, context -> {
+      ProvisioningTaskInfo currentTaskInfo = getProvisionerTable(context).getTaskInfo(taskKey);
+      if (currentTaskInfo == null) {
+        return null;
+      }
+      // write that the state has been cancelled. This is in case CDAP dies or is killed before the cluster can
+      // be deprovisioned and the task state cleaned up. When CDAP starts back up, it will see that the task is
+      // cancelled and will not resume the task.
+      ProvisioningOp newOp =
+        new ProvisioningOp(currentTaskInfo.getProvisioningOp().getType(), ProvisioningOp.Status.CANCELLED);
+      ProvisioningTaskInfo newTaskInfo = new ProvisioningTaskInfo(currentTaskInfo, newOp, currentTaskInfo.getCluster());
+      getProvisionerTable(context).putTaskInfo(newTaskInfo);
```

Comment:
It just came to me that this may not be safe to do. Does the context cache what you have already instantiated? If not, then you would have two instances of the same table, participating in the same transaction. Because our HBase tables buffer changes in-memory, they may overwrite each other's changes. 
I think it is safe in this case, because the first instance is only used for reading. But I think it would be better to acquire the table once per transaction. 
At the same time, we need to validate and add possibly fix the StructuredTableContext to make sure that repeated get() calls for the same table do not return multiple instances. Previously, we had the DynamicDatasetCache for that, but we are now not using it.

Reply:
Updated. Also updated copyright for this file. Will squash and merge since we don't need another review